### ISSUE TITLE
zero row query issue with SetTableFormat

### ISSUE THREAD

**Issue by lee-lindley** [mbleron/ExcelGen#15]
When creating a sheet from a query or cursor that returns 0 rows, and also setting a style using setTableFormat, Excel complains and offers to recover the workbook. If you answer Yes, the sheet with the column headers only is provided as expected, but without the style
![Screenshot Excel Error Dialog](https://user-images.githubusercontent.com/87572140/152453771-6e42b9f7-35b4-48cc-ac34-cc416dc60789.gif)
.
If I comment out setTableFormat on line 11, Excel does not complain.

Workaround is to determine whether query will return any rows and only set the style if it does. Unfortunately, ExcelGen does not expose back to the caller the number of rows returned by the query or on the sheet. Whether the specific issue with the style can be addressed or not, it would be useful to add a function to return the number of rows the query generated for any given sheet.

Test case code:
```sql
WITH 
FUNCTION get_xlsx RETURN BLOB AS
    v_blob          BLOB;
    v_ctxId         ExcelGen.ctxHandle;
    v_sheetHandle   BINARY_INTEGER;
BEGIN
        v_ctxId := ExcelGen.createContext();
        v_sheetHandle := ExcelGen.addSheetFromQuery(v_ctxId, 'Test 0 rows','SELECT SYSDATE AS "my date col" FROM DUAL WHERE 1=0', p_sheetIndex => 1);
        -- freeze the top row with the column headers
        ExcelGen.setHeader(v_ctxId, v_sheetHandle, p_frozen => TRUE);
        ExcelGen.SetTableFormat(p_ctxid=> v_ctxId, p_sheetid=> v_sheetHandle, p_style=> 'TableStyleMedium7');
        
        v_blob := ExcelGen.getFileContent(v_ctxId);
        ExcelGen.closeContext(v_ctxId);
        RETURN v_blob;
END;
SELECT get_xlsx FROM DUAL
;
```

**Comment by lee-lindley** [comment]
On second thought, I'm not sure the query is executed until after we must set the style, so it may not be helpful to get the number of rows.

**Comment by mbleron** [comment]
Hi Lee,

> On second thought, I'm not sure the query is executed until after we must set the style, so it may not be helpful to get the number of rows.  

Exactly, `addSheetFromQuery` is only declarative. All queries are executed when calling `getFileContent`.

I'll look into the style issue this weekend.
Thanks for reporting it.

**Comment by mbleron** [comment]
Fixed in v2.3.
For an empty dataset, table formatting will _visually_ appear like this, with a mandatory empty row : 
![image](https://user-images.githubusercontent.com/19926388/152690866-8bcc4730-907b-4c67-ab16-fbe3cb188659.png)

I've implemented that by extending the table range by one row down, but that row doesn't exist internally so the actual sheet content is correct.

**Comment by lee-lindley** [comment]
Awesome. Thanks Marc.
 
I understand in the case of 0 rows returned, we produce an empty row. When there are 1 or more rows returned, we do not. On a normal sheet when one does <CTRL> <down arrow> Excel puts the cursor in the last populated row. On an empty query sheet, it goes into the empty row.
 
I believe this will make my team happy. I'm sure somebody will harp about that empty row, but I'll remind them they can always run a check to make sure a query returns rows if they want to work around it. I can't imagine our business team complaining about the empty row. If they do, I'll offer to go back to individual CSV files. Should silence the peanut gallery.
 
-- 
Lee Lindley
***@***.***
